### DISCUSSION TITLE
names with '-' and map with constraints

### DISCUSSION THREAD

**Discussion by brooksgx** [cue-lang/cue#1377] [Q&A] [answered]
Hi,

I am using the go integration for a CLI to validate some YAML.

I would like to be able to define name with '-' as a preference to underscores or mixed case names

e.g.

`my-compound-name: string`

Is there a way to escape the '-' or force the name to be accepted as defined?

Objects as lists:

As part of the same schema I am trying to define key:value constructs with some constraints on the value. In this case a map of string to integer as a list.

```
object-as-list:
  "my key": 1
  "my second key": 2

```
Is there a way to model this in Cue and verify that the values are integer values?

Thanks in advance!!

**Comment by seh** [comment, marked as answer]
Does [this example in the playground](https://cuelang.org/play/?id=DDEPtZzXvgx#cue@export@yaml) accomplish what you're trying to express?

**Reply by brooksgx** [reply]
Two questions answered with one example. Thanks that helps a lot

**Reply by seh** [reply]
Note that if you wanted to enforce that your key names are only alphabetic with intervening hyphens, you could write that constraint, per [this updated example](https://cuelang.org/play/?id=PmfxubCKSIH#cue@export@cue).

Season to taste.

**Reply by brooksgx** [reply]
That's even more helpful at being able to restrict the keys in the map to particular patterns.

**Reply by verdverm** [reply]
You can use the regexp directly in the pattern constraints with a definition to prevent additional fields

```cue
#KeyFmt: =~"^[[:lower:]]+(?:-[[:lower:]]+)*$"

#ol: [#KeyFmt]: int

"object-as-list": #ol & {
  "my-key": 1
  "my-second-key": 2
}
```

https://cuelang.org/play/?id=D_3FHu0iI8W#cue@export@cue

**Reply by seh** [reply]
I kept trying to write [the following](https://cuelang.org/play/?id=1MtXyK73Z75#cue@export@cue), and couldn't get it to work:
```cue
"object-as-list": [_#KeyFmt]: int // Assume the definition _#KeyFmt exists.
```
I couldn't get that to catch any violations in the key names.

**Reply by verdverm** [reply]
You can also use `close` to get the same effect

https://cuelang.org/play/?id=FvX6Fb02BWL#cue@export@cue

**Reply by seh** [reply]
Ah, nice. Thank you, Tony.

**Reply by brooksgx** [reply]
Great examples - incorporated into the schema I'm working on as part of the validation steps for data.
